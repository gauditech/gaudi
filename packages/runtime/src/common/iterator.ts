export type Iterator = {
  /** Value of current iteration */
  current: number;
  /** Total number of possible iterations */
  total: number;
};

/**
 * Create iterator generator fn which returns `Iterator` with every iteration.
 *
 * This is a JS generator function which means it can be directly used with
 * `for of`, `forEach` and similar iteration constructs
 *
 * Eg.
 * ```
 * for(let iter of createIterator(1, 5)) {
 *   console.log("Value:" + iter.current)
 * }
 * ```
 */
export function* createIterator(start: number, end: number): Generator<Iterator, void, void> {
  let current = start;
  const total = end - start + 1;

  for (let i = 0; i < total; i++) {
    yield { current, total };
    current++;
  }
}
