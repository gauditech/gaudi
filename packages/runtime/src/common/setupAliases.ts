import path from "path";

import { addAliases } from "module-alias";

/**
 * NOTE: Make sure to keep aliases in sync with `tsconfig.json` and `jest.config.js`.
 */

addAliases({
  "@runtime": path.join(__dirname, "src"),
});
