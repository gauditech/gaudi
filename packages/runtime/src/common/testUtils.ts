import { compileToAST } from "@gaudi/compiler/dist/compiler";
import { compilerErrorsToString } from "@gaudi/compiler/dist/compiler/compilerError";
import { migrate } from "@gaudi/compiler/dist/compiler/migrate";
import { compose } from "@gaudi/compiler/dist/composer/composer";
import {
  CustomManyEndpointDef,
  Definition,
  QueryAction,
  QueryDef,
} from "@gaudi/compiler/dist/types/definition";

import { QueryTree } from "@runtime/query/build";
import { NestedRow, QueryExecutor } from "@runtime/query/exec";
import { Storage } from "@runtime/server/context";

/**
 * Helper function that compiles directly to definition. This is used in tests.
 */
export function compileFromString(source: string): Definition {
  const inputs = [{ source }];
  const { ast, errors } = compileToAST([{ source }]);
  if (errors) {
    const errorString = compilerErrorsToString(inputs, errors);
    throw new Error(`Failed to compile:\n${errorString}`);
  }
  return compose(migrate(ast));
}

/**
 * Creates dummy query executor wich always return empty row.
 *
 * TODO: add abbility to define dummy return data
 */
export function mockQueryExecutor(): QueryExecutor {
  return {
    executeQuery(
      _def: Definition,
      _q: QueryDef,
      _ctx: Storage,
      _contextIds: number[]
    ): Promise<NestedRow[]> {
      return Promise.resolve([]);
    },

    executeQueryTree(
      _def: Definition,
      _qt: QueryTree,
      _ctx: Storage,
      _contextIds: number[]
    ): Promise<NestedRow[]> {
      return Promise.resolve([]);
    },
  };
}
/**
 * Provide models blueprint, and query blueprint separately.
 * This function will compile a blueprint and extract the query definition,
 * so you can use it in tests.
 */

export function makeTestQuery(models: string, query: string): { def: Definition; query: QueryDef } {
  const bp = `
  ${models}

  model TestHelperModel {}
  api {
    entrypoint TestHelperModel {
      custom endpoint {
        method GET
        cardinality many
        path "/test"
        action {
          ${query}
        }
      }
    }
  }
  `;
  const def = compileFromString(bp);
  const endpoint = def.apis[0].entrypoints[0].endpoints[0] as CustomManyEndpointDef;
  const action = endpoint.actions[0] as QueryAction;
  return { def, query: action.query };
}
