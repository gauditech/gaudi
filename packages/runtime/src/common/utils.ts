import flat from "flat";
import _ from "lodash";

/** Calculate query limit/offset from page/pageSize. */
export function pagingToQueryLimit(
  page: number | undefined,
  pageSize: number | undefined,
  defaultOffset: number | undefined,
  defaultLimit: number | undefined
): { limit: number | undefined; offset: number | undefined } {
  let limit = defaultLimit;
  if (pageSize != null) {
    // make sure it's >= 0, if it's negative, default to 0
    limit = Math.max(pageSize, 0);
  }

  let offset = defaultOffset;
  if (page != null && limit != null) {
    // "page" is 1-based, offset is 0-based
    // make sure it's >= 0
    offset = Math.max(page - 1, 0) * limit;
  }

  return { limit, offset };
}

/** Collect deeply-nested structures into a flat array. */
export function collect(values: any, path: string[]): unknown[] {
  if (_.isEmpty(path)) {
    return values;
  }
  const [name, ...rest] = path;
  if (_.isArray(values)) {
    return _.compact(values.flatMap((v) => collect(_.get(v, name), rest)));
  } else {
    return collect(_.get(values, name), rest);
  }
}

/**
 * Flattens a nested object/record.
 * Keys are created recursively, using __ as delimiter.
 * */
export function flatten(record: Record<string, any>): Record<string, unknown> {
  return flat.flatten(_.omit(record, "_server"), { delimiter: "__" });
}
