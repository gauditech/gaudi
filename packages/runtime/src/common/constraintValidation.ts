import _ from "lodash";
import { match } from "ts-pattern";

import { GAUDI_INTERNAL_TARGET_ID_ALIAS, queryFromParts, selectableId } from "../query/build";
import { executeQuery } from "../query/exec";
import { DbConn } from "../server/dbConn";
import { Vars } from "../server/vars";

import { getRef } from "@gaudi/compiler/common/refs";
import { getTypedPathWithLeaf } from "@gaudi/compiler/composer/utils";
import {
  ActionDef,
  Definition,
  FieldsetDef,
  FieldsetFieldDef,
  TypedExprDef,
} from "@gaudi/compiler/types/definition";

export type ReferenceIdResult =
  | ValidReferenceIdResult
  | InvalidReferenceIdResult
  | InputMissingReferenceIdResult;

export type ValidReferenceIdResult = {
  kind: "reference-found";
  fieldsetAccess: string[];
  value: number;
};
type InvalidReferenceIdResult = {
  kind: "reference-not-found";
  fieldsetAccess: string[];
};
type InputMissingReferenceIdResult = {
  kind: "reference-input-missing";
  fieldsetAccess: string[];
  inputValue: null | undefined;
};

export async function fetchReferenceIds(
  def: Definition,
  dbConn: DbConn,
  actions: ActionDef[],
  input: Record<string, unknown>
): Promise<ReferenceIdResult[]> {
  const referenceInputs = actions.flatMap((action) => {
    if (action.kind !== "create-one" && action.kind !== "update-one") return [];
    return action.changeset.flatMap(({ name, setter }) => {
      if (setter.kind !== "fieldset-reference-input") return [];
      return [[action.model, name, setter] as const];
    });
  });

  const promiseEntries = referenceInputs.map(
    async ([model, name, setter]): Promise<ReferenceIdResult> => {
      const inputValue = _.get(input, setter.fieldsetAccess);
      const { fieldsetAccess } = setter;
      if (_.isNil(inputValue)) {
        return {
          kind: "reference-input-missing",
          fieldsetAccess,
          inputValue,
        };
      }
      const reference = getRef.reference(def, model, name);
      const tpath = getTypedPathWithLeaf(def, [reference.toModelRefKey, ...setter.through], {});
      const field = getRef.field(def, tpath.leaf.refKey);

      const varName = [...tpath.fullPath, "_input"].join("_");
      const filter: TypedExprDef = {
        kind: "function",
        name: "is",
        args: [
          {
            kind: "alias",
            namePath: tpath.fullPath,
          },
          {
            kind: "variable",
            name: varName,
            type: { kind: field.type, nullable: field.nullable },
          },
        ],
      };
      const queryName = tpath.fullPath.join(".");
      const query = queryFromParts(def, queryName, _.initial(tpath.fullPath), filter, [
        selectableId(_.dropRight(tpath.fullPath, 2)),
      ]);
      const result = await executeQuery(
        dbConn,
        def,
        query,
        new Vars({ [varName]: inputValue }),
        []
      );

      if (result.length === 0) {
        return { kind: "reference-not-found", fieldsetAccess };
      }
      if (result.length > 1) {
        throw Error(
          `Failed to find reference: There are multiple (${result.length}) '${field.modelRefKey}' where field '${field.name}' is '${inputValue}'`
        );
      }

      return {
        kind: "reference-found",
        fieldsetAccess: setter.fieldsetAccess,
        value: result[0][GAUDI_INTERNAL_TARGET_ID_ALIAS] as number,
      };
    }
  );
  return Promise.all(promiseEntries);
}

/**
 * This mutates endpoint fieldset.
 * If reference has no id it will add "noReference" validator to fieldset field.
 */
export function assignNoReferenceValidators(
  fieldset: FieldsetDef,
  referenceIds: ReferenceIdResult[]
): asserts referenceIds is ValidReferenceIdResult[] {
  referenceIds.forEach((referenceIdResult) => {
    match(referenceIdResult)
      .with({ kind: "reference-not-found" }, ({ fieldsetAccess }) => {
        const currentFieldset = getNestedFieldset(fieldset, fieldsetAccess);
        currentFieldset.referenceNotFound = true;
      })
      .otherwise(_.noop);
  });
}

function getNestedFieldset(fieldset: FieldsetDef, fieldsetAccess: string[]): FieldsetFieldDef {
  if (fieldsetAccess.length === 0) {
    if (fieldset.kind === "record") {
      throw Error("Unexpected FieldsetRecordDef when searching for a nested fieldset");
    }
    return fieldset;
  }
  if (fieldset.kind === "field") {
    throw Error("Unexpected FieldsetFieldDef when searching for a nested fieldset");
  }
  const [head, ...tail] = fieldsetAccess;
  return getNestedFieldset(fieldset.record[head], tail);
}
