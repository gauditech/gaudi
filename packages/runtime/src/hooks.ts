import { promises as fs } from "fs";
import path from "path";

import { Request, Response } from "express";

import { getExecutionRuntime } from "@gaudi/compiler/dist/common/refs";
import { getInternalExecutionRuntimeName } from "@gaudi/compiler/dist/composer/executionRuntimes";
import { HookCode, HookInline, HookSource } from "@gaudi/compiler/dist/types/common";
import { Definition, ExecutionRuntimeDef } from "@gaudi/compiler/dist/types/definition";

const EXECUTION_RUNTIMES: Record<string, ExecutionRuntimeClient> = {};

/**
 * Executes hook.
 *
 * This is used in model/validator/setter hooks which all work
 * with gaudi-controlled arguments.
 */
export async function executeHook<T>(
  def: Definition,
  hook: HookCode,
  args: Record<string, unknown>
): Promise<T> {
  if (hook.kind === "inline") return executeInlineHook(hook, args);
  const execRuntime = getExecutionRuntime(def, hook.runtimeName);
  return (await createExecutionRuntime(execRuntime)).executeHook<T>(hook, args);
}

/**
 * Executes action hook.
 *
 * Action hooks are used to provide action implementation. They can receive gaudi-controlled
 * arguments but also, they need access to low level abstractions and that's why they
 * receive other objects like eg. request/response.
 */
export async function executeActionHook<T>(
  def: Definition,
  hook: HookCode,
  args: Record<string, unknown>,
  ctx: {
    request: Request;
    response: Response;
  }
): Promise<T> {
  if (hook.kind === "inline") return executeInlineHook(hook, args);
  const execRuntime = getExecutionRuntime(def, hook.runtimeName);
  return (await createExecutionRuntime(execRuntime)).executeHook<T>(hook, args, ctx);
}

/**
 * Executes inline hook.
 *
 * Inline hook is seperate because it can be executed without named runtime environment.
 */
export async function executeInlineHook<T>(
  hook: HookInline,
  args: Record<string, unknown>
): Promise<T> {
  // order of args must be consistent
  const argNames = Object.entries(args).map(([name, _value]) => name);
  const argValues = Object.entries(args).map(([_name, value]) => value);
  const hookBody = [
    // strict mode
    // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Strict_mode
    "'use strict'",
    // hook fn body
    `return ${hook.inline}`,
  ].join("\n");

  // dynamically create new function
  // allows access only to global and function's own scope
  // syntax is `new AsyncFunction(arg1, ..., argN, fnBody)`
  // arguments are positional so order in fn definition (here) and in fn call (below) must be the same
  const hookFn = new AsyncFunction(...[...argNames, hookBody]);

  // TODO: cache inline function - by which key, source code?!

  return hookFn(...argValues);
}

async function createExecutionRuntime(
  runtime: ExecutionRuntimeDef
): Promise<ExecutionRuntimeClient> {
  if (EXECUTION_RUNTIMES[runtime.name] == null) {
    // only local gaudi execution runtime is supported at the moment
    EXECUTION_RUNTIMES[runtime.name] = await createLocalExecutionRuntime(runtime);
  }

  return EXECUTION_RUNTIMES[runtime.name];
}

// ---------- Local execution runtime

export type HookActionContext = { request: Request; response: Response };
export type HookFunction = <T>(
  args: Record<string, unknown>,
  ctx?: HookActionContext
) => Promise<T>;
export type HookModules = Record<string, Record<string, HookFunction>>;

export type ExecutionRuntimeClient = {
  runtimeName: string;
  executeHook: <T>(
    hook: HookSource,
    args: Record<string, unknown>,
    ctx?: HookActionContext
  ) => Promise<T>;
};

const HOOKS_FILES_PATTERN = /.+\.[tj]s$/;

export async function createLocalExecutionRuntime(
  runtime: ExecutionRuntimeDef
): Promise<ExecutionRuntimeClient> {
  const modules: HookModules = {};

  const sourcePath = resolveSourcePath(runtime.name, runtime.sourcePath);
  await importHooks(sourcePath, modules);

  return {
    runtimeName: runtime.name,
    executeHook(hook, args, ctx) {
      const hookFile = modules[hook.file];
      if (hookFile == null) {
        throw new Error(`Hook file "${hook.file}" not found`);
      }

      const hookFn = hookFile[hook.target];
      if (hookFn == null) {
        throw new Error(`Hook "${hook.target}" in file "${hook.file}" not found`);
      }

      return hookFn(args, ctx);
    },
  };
}

/**
 * Resolves absolute source folder path.
 *
 * If sources path is already absolute it is returned as umchanged.
 *
 * Relative paths are resolved with different base path:
 *  - internal exec runtimes are resolved using `__dirname` as base path
 *  - user defined exec runtimes are resolved using current process working folder
 */
function resolveSourcePath(name: string, sourcePath: string): string {
  if (path.isAbsolute(sourcePath)) return sourcePath;

  const internalExecRuntimes = [getInternalExecutionRuntimeName()];
  if (internalExecRuntimes.includes(name)) {
    return path.resolve(__dirname, sourcePath);
  } else {
    return path.resolve(sourcePath);
  }
}

async function importHooks(sourcePath: string, modules: HookModules) {
  console.log("Loading hooks sources from:", path.resolve(sourcePath));

  async function loadHooksFromDir(dir: string) {
    const entities = await fs.readdir(path.join(sourcePath, dir));

    const promises = entities.map(async (entityFilename) => {
      const hookPath = path.join(dir, entityFilename);
      const entity = path.join(sourcePath, hookPath);

      const stats = await fs.lstat(entity);

      if (stats.isDirectory()) {
        await loadHooksFromDir(hookPath);
      } else if (stats.isFile() && HOOKS_FILES_PATTERN.test(entityFilename)) {
        modules[hookPath] = loadFileAsModule(entity);
      }
    });

    await Promise.all(promises);
  }

  await loadHooksFromDir("");
}

function loadFileAsModule(filepath: string) {
  const absolute = path.resolve(filepath);
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  return require(absolute);
}

// we cannot dynamically create async function using `new Function()` so we'll hijack real async function's prototype to create a new one dynamically
const AsyncFunction = Object.getPrototypeOf(async function () {
  // empty body
}).constructor;
