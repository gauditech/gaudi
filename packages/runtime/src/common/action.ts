import { transformSelectPath } from "@gaudi/compiler/dist/common/query";
import { dataToFieldDbnames, getRef } from "@gaudi/compiler/dist/common/refs";
import { assertUnreachable, ensureExists } from "@gaudi/compiler/dist/common/utils";
import {
  ActionDef,
  CreateOneAction,
  Definition,
  UpdateOneAction,
} from "@gaudi/compiler/dist/types/definition";
import _ from "lodash";

import { applyFilterIdInContext, buildQueryTree, queryTreeFromParts } from "../query/build";
import { castToCardinality, createQueryExecutor, executeQueryTree } from "../query/exec";

import { ValidReferenceIdResult } from "./constraintValidation";

import { buildChangeset } from "@runtime/common/changeset";
import { HookActionContext, executeActionHook } from "@runtime/hooks";
import { DbConn } from "@runtime/server/dbConn";
import { HookError } from "@runtime/server/error";
import { Vars } from "@runtime/server/vars";

export type ActionContext = {
  input: Record<string, unknown>;
  vars: Vars;
  referenceIds: ValidReferenceIdResult[];
};

export async function executeEndpointActions(
  def: Definition,
  dbConn: DbConn,
  ctx: ActionContext,
  epCtx: HookActionContext,
  actions: ActionDef[]
) {
  //
  return _internalExecuteActions(def, dbConn, ctx, epCtx, actions);
}
export async function executeActions(
  def: Definition,
  dbConn: DbConn,
  ctx: ActionContext,
  actions: ActionDef[]
) {
  return _internalExecuteActions(def, dbConn, ctx, undefined, actions);
}
async function _internalExecuteActions(
  def: Definition,
  dbConn: DbConn,
  ctx: ActionContext,
  epCtx: HookActionContext | undefined,
  actions: ActionDef[]
) {
  const qx = createQueryExecutor(dbConn);

  for (const action of actions) {
    const actionKind = action.kind;
    if (actionKind === "create-one") {
      const model = getRef.model(def, action.model);
      const dbModel = model.dbname;

      const actionChangeset = await buildChangeset(def, qx, epCtx, action.changeset, ctx);
      const dbData = dataToFieldDbnames(model, actionChangeset);

      const id = await insertData(dbConn, dbModel, dbData);
      const deps = await fetchActionDeps(def, dbConn, action, id);
      deps && ctx.vars.set(action.alias, deps[0]);
    } else if (actionKind === "update-one") {
      const model = getRef.model(def, action.model);
      const dbModel = model.dbname;

      const actionChangeset = await buildChangeset(def, qx, epCtx, action.changeset, ctx);
      const dbData = dataToFieldDbnames(model, actionChangeset);

      const targetId = resolveTargetId(ctx, action.targetPath);

      const id = await updateData(dbConn, dbModel, dbData, targetId);
      const deps = await fetchActionDeps(def, dbConn, action, id);
      deps && ctx.vars.set(action.alias, deps[0]);
    } else if (actionKind === "delete-one") {
      const model = getRef.model(def, action.model);
      const dbModel = model.dbname;

      const targetId = resolveTargetId(ctx, action.targetPath);

      await deleteData(dbConn, dbModel, targetId);
    } else if (actionKind === "query") {
      const qt = buildQueryTree(def, action.query);
      // FIXME this ugly
      const varsObj = Object.fromEntries(
        Object.keys(ctx.input).map((k) => [`___changeset___${k}`, ctx.input[k]])
      );
      varsObj["___requestAuthToken"] = epCtx?.request.user?.token;

      const result = castToCardinality(
        await qx.executeQueryTree(def, qt, new Vars(varsObj), []),
        action.query.retCardinality
      );
      ctx.vars.set(action.alias, result);
    } else if (actionKind === "execute-hook") {
      ensureExists(epCtx, '"execute" actions can run only in endpoint context.');

      const argsChangeset = await buildChangeset(def, qx, epCtx, action.hook.args, ctx);

      try {
        const result = await executeActionHook(def, action.hook.hook, argsChangeset, epCtx);
        ctx.vars.set(action.alias, result);
      } catch (err) {
        throw new HookError(err);
      }
    } else if (actionKind == "respond") {
      ensureExists(epCtx, '"execute" actions can run only in endpoint context.');

      try {
        // construct changeset and resolve action parts
        const actionChangesetDef = _.compact([
          { name: "body", setter: action.body },
          action.httpStatus != null ? { name: "httpStatus", setter: action.httpStatus } : undefined,
        ]);
        const changeset = await buildChangeset(def, qx, epCtx, actionChangesetDef, ctx);
        // http headers are a separate changeset
        const httpHeadersChangesetDef = (action.httpHeaders ?? []).map((h) => ({
          name: h.name,
          setter: h.value,
        }));
        const httpHeadersChangeset = await buildChangeset(
          def,
          qx,
          epCtx,
          httpHeadersChangesetDef,
          ctx
        );

        const body = changeset.body;
        // we're forcing number cause our type system should've made sure that this resolves to appropriate type
        const httpResponseCode = (changeset.httpStatus ?? 200) as number;

        Object.entries(httpHeadersChangeset).forEach(([name, value]) => {
          // null - remove current header
          if (value == null) {
            epCtx.response.removeHeader(name);
          }
          // multiple header values
          else if (_.isArray(value)) {
            epCtx.response.set(name, value);
          }
          // single value
          else {
            // we're forcing value to `any` cause our type system should've made sure that this resolves to appropriate type
            epCtx.response.set(name, value as any);
          }
        });
        epCtx.response.status(httpResponseCode).json(body);
      } catch (err: any) {
        throw new Error(err);
      }
    } else if (actionKind === "validate") {
      throw new Error("Not implemented");
    } else {
      assertUnreachable(actionKind);
    }
  }
}

/**
 * Fetches the record targeted by action based on the actions `select` deps.
 */
async function fetchActionDeps(
  def: Definition,
  dbConn: DbConn,
  action: CreateOneAction | UpdateOneAction,
  id: number
): Promise<Record<string, unknown>[] | undefined> {
  if (!action.alias) return;
  // no need to fetch if only ID is requested
  if (action.select.findIndex((item) => item.alias !== "id") < 0) {
    return [{ id }];
  }
  const qt = queryTreeFromParts(
    def,
    action.alias,
    [action.model],
    applyFilterIdInContext([action.model]),
    transformSelectPath(action.select, [action.alias], [action.model])
  );
  return executeQueryTree(dbConn, def, qt, new Vars(), [id]);
}

function resolveTargetId(ctx: ActionContext, targetPath: string[]): number {
  // append "id" to the end of the target path and separate it from it's root
  // eg.
  // [org, repo, issue] -> [org, repo, issue, id] -> [org, [repo, issue, id]]
  const [rootTargetName, ...path] = [...targetPath, "id"];

  return ctx.vars.get(rootTargetName, path);
}

// ---------- DB functions

async function updateData(
  dbConn: DbConn,
  model: string,
  data: Record<string, unknown>,
  targetId: number
): Promise<number> {
  // avoid malformed query by skipping the update if no data is passed
  if (Object.keys(data).length === 0) {
    return targetId;
  }
  const ret = await dbConn(model).update(data).where({ id: targetId }).returning("id");
  return findOne(ret);
}

async function insertData(
  dbConn: DbConn,
  model: string,
  data: Record<string, unknown>
): Promise<number> {
  const ret = await dbConn.insert(data).into(model).returning("id");

  return findOne(ret);
}

async function deleteData(dbConn: DbConn, model: string, targetId: number): Promise<number> {
  const ret = await dbConn(model).delete().where({ id: targetId }).returning("id");

  return findOne(ret);
}

function findOne(rows: any[], message?: string) {
  if (rows.length === 0) {
    throw new Error(message ?? `Record not found`);
  }
  if (rows.length > 1) {
    throw new Error(`Unexpected error: multiple records found`);
  }

  return rows[0].id;
}
