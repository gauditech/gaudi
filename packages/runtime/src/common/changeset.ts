import _, { get, indexOf, isString, set, toInteger, toString } from "lodash";

import { buildQueryTree } from "../query/build";

import { executeArithmetics } from "./arithmetics";

import {
  assertUnreachable,
  ensureEqual,
  ensureExists,
  ensureNot,
} from "@gaudi/compiler/common/utils";
import { ActionContext } from "@runtime/common/action";
import { HookActionContext, executeHook } from "@runtime/hooks";
import { QueryExecutor, castToCardinality } from "@runtime/query/exec";
import { ChangesetDef, Definition, FieldDef, FieldSetter } from "@gaudi/compiler/types/definition";

type Changeset = Record<string, unknown>;

/**
 * Build result record from given action changeset rules and give context (source) inputs.
 */
export async function buildChangeset(
  def: Definition,
  qx: QueryExecutor,
  epCtx: HookActionContext | undefined,
  actionChangsetDefinition: ChangesetDef,
  actionContext: ActionContext,
  // `changesetContext` is used for hooks, to be able to pass the "parent context" changeset
  changesetContext: Changeset = {}
): Promise<Changeset> {
  const changeset: Changeset = {};

  async function getValue(setter: FieldSetter): Promise<unknown> {
    switch (setter.kind) {
      case "literal": {
        return formatFieldValue(setter.literal.value, setter.literal.kind);
      }
      case "fieldset-input": {
        return formatFieldValue(
          getFieldsetProperty(actionContext.input, setter.fieldsetAccess),
          setter.type
        );
      }
      case "reference-value": {
        return actionContext.vars.collect([setter.target.alias, ...setter.target.access]);
      }
      case "fieldset-hook": {
        const args = await buildChangeset(def, qx, epCtx, setter.args, actionContext, changeset);
        return await executeHook(def, setter.hook, args);
      }
      case "changeset-reference": {
        /**
         * Composer guarantees the correct order of array elements, so `setter.referenceName` should
         * be in the context and we should be able to return `changeset[setter.referenceName]`.
         *
         * However, hooks inherit the action changeset context, but also build their own changeset,
         * so we need to check which changeset `setter.referenceName` belongs to. In other words,
         * it's possible that `setter.referenceName in changeset` is `false` when building a hooks changeset,
         * but in that case we can assume it's in the `changesetContext`.
         *
         * NOTE the following code wouldn't work because `undefined` is a valid changeset value:
         * `return changeset[setter.referenceName] || changesetContext[setter.referenceName];`
         */
        if (setter.referenceName in changeset) {
          return changeset[setter.referenceName];
        } else {
          ensureEqual(
            setter.referenceName in changesetContext,
            true,
            `Reference "${setter.referenceName}" not found in changeset context`
          );
          return changesetContext[setter.referenceName];
        }
      }
      case "fieldset-reference-input": {
        const referenceIdResult = actionContext.referenceIds.find((result) =>
          _.isEqual(result.fieldsetAccess, setter.fieldsetAccess)
        );
        ensureNot(
          referenceIdResult,
          undefined,
          `${JSON.stringify(actionContext.referenceIds)} -> ${setter.fieldsetAccess}`
        );
        return referenceIdResult.value;
      }
      case "request-auth-token": {
        ensureExists(epCtx, `HTTP handle context is required for "${setter.kind}" changesets`);
        const handler = epCtx.request;

        return getFieldsetProperty(handler, setter.access);
      }
      case "function": {
        return executeArithmetics(setter, (s) => getValue(s));
      }
      case "context-reference": {
        return actionContext.vars.get(setter.referenceName);
      }
      case "query": {
        const vars = actionContext.vars.copy();
        vars.set("___requestAuthToken", await getValue({ kind: "request-auth-token", access: [] }));
        Object.keys(actionContext.input).forEach((key) => {
          vars.set(`___changeset___${key}`, actionContext.input[key]);
        });

        const qt = buildQueryTree(def, setter.query);
        const results = await qx.executeQueryTree(def, qt, vars, []);
        return castToCardinality(results, setter.query.retCardinality);
      }
      case "array": {
        return await Promise.all(setter.elements.map((e) => getValue(e)));
      }
      default: {
        return assertUnreachable(setter);
      }
    }
  }

  for (const { name, setter } of actionChangsetDefinition) {
    switch (setter.kind) {
      case "fieldset-reference-input": {
        changeset[name + "_id"] = await getValue(setter);
        break;
      }
      default: {
        changeset[name] = await getValue(setter);
      }
    }
  }

  return changeset;
}

/**
 * Format unknown field value to one of mapping types.
 *
 * If `value` is `undefined`/`null`, an `undefined` is returned.
 *
 * This function uses `lodash` to convert "string" and "integer".
 * For details see: https://lodash.com/docs
 *
 * Mappings:
 * - string - `string`, `_.toString()`
 * - integer - `number`, `_.toInteger()`
 * - boolean - `boolean`, see below
 *
 * Due to `lodash` lacking boolean converter, this fn does it manually following these rules:
 * - "true" - string "true", case insensitive ("true", "TRUE", "TruE", ...)
 * - true - real boolean true
 * - everything else is converted to `false`
 *
 * TODO: move to some utils folder if it's ok
 */
export function formatFieldValue(
  value: unknown,
  type: FieldDef["type"] | "null"
): string | number | boolean | undefined | null {
  if (_.isNil(value)) return value;

  if (type === "boolean") {
    if (isString(value)) {
      value = value.toLowerCase();
    }

    return indexOf(["true", true], value) != -1;
  } else if (type === "integer") {
    return toInteger(value);
  } else {
    return toString(value);
  }
}

// ----- transformations

export function getFieldsetProperty<T = unknown>(target: unknown, fieldsetAccess: string[]): T {
  return get(target, fieldsetAccessToPath(fieldsetAccess));
}
export function setFieldsetProperty(
  target: object,
  fieldsetAccess: string[],
  value: unknown
): unknown {
  set(target, fieldsetAccessToPath(fieldsetAccess), value);

  return target;
}

export function fieldsetAccessToPath(access: string[]): string {
  return access.join(".");
}
