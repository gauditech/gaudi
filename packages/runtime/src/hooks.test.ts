import { kindFilter } from "@gaudi/compiler/common/kindFilter";
import { getInternalExecutionRuntimeName } from "@gaudi/compiler/composer/executionRuntimes";
import { compileFromString } from "@runtime/common/testUtils";
import { executeHook } from "@runtime/hooks";
import { Definition } from "@gaudi/compiler/types/definition";

describe("hooks", () => {
  const def = createTestDefinition();

  /*
   * Using math division as a test operation:
   *  - tests multiple arguments
   *  - division is not cummutative operation and will detect errors if arguments are not passed in a specific order
   */

  describe("external hooks", () => {
    it("should resolve static value", async () => {
      const result = await executeHook(
        def,
        { kind: "source", target: "divideStatic", file: "hooks.js", runtimeName: "TestRuntime" },
        { x: 6, y: 2 }
      );

      expect(result).toBe(3);
    });

    it("should resolve promise value", async () => {
      const result = await executeHook(
        def,
        { kind: "source", target: "divideAsync", file: "hooks.js", runtimeName: "TestRuntime" },
        { x: 6, y: 2 }
      );

      expect(result).toBe(3);
    });
  });

  describe("inline hooks", () => {
    it("should resolve static value", async () => {
      const result = await executeHook(def, { kind: "inline", inline: "x / y" }, { x: 6, y: 2 });

      expect(result).toBe(3);
    });

    it("should resolve promise value", async () => {
      const result = await executeHook(
        def,
        { kind: "inline", inline: "Promise.resolve(x / y)" },
        { x: 6, y: 2 }
      );

      expect(result).toBe(3);
    });
  });

  describe("execution runtime", () => {
    it("should execute hooks from multiple exec runtimes", async () => {
      const bp = `
      runtime MathRuntime {
        default
        source path "./src/runtime/test/hooks"
      }

      runtime TextRuntime {
        source path "./src/runtime/test/hooks2"
      }

      model Result {
        field name { type string }
        field avg { type integer }
      }

      api {
        entrypoint Result {
          create endpoint {
            action {
              create {
                set name hook {
                  runtime TextRuntime
                  arg value "First Last"
                  arg prefix "Mr. "
                  source prefix from "hooks2.js"
                }
                set avg hook {
                  // thisi is from the default runtime
                  arg x 100
                  arg y 20
                  source prefix from "hooks.js"
                }
              }
            }
          }
        }
      }

    `;

      const result = compileFromString(bp);
      const action = kindFilter(result.apis[0].entrypoints[0].endpoints, "create")
        .shift()
        ?.actions.at(0);

      expect(action).toMatchSnapshot();
    });

    it("should run hook from internal exec runtime", async () => {
      const def = createTestDefinition();

      const result = await executeHook(
        def,
        {
          kind: "source",
          file: "hooks/index.js",
          target: "echo",
          runtimeName: getInternalExecutionRuntimeName(),
        },
        { value: "ASDF" }
      );

      expect(result).toBe("ASDF");
    });
  });
});

/**
 * Creates test definition struct
 */
function createTestDefinition(): Definition {
  const bp = `
    runtime TestRuntime {
      default
      source path "./src/runtime/test/hooks"
    }
  `;
  return compileFromString(bp);
}
