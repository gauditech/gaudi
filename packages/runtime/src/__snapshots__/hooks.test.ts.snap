// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hooks execution runtime should execute hooks from multiple exec runtimes 1`] = `
{
  "alias": "$action_0",
  "changeset": [
    {
      "kind": "basic",
      "name": "name",
      "setter": {
        "hook": {
          "args": [
            {
              "kind": "basic",
              "name": "value",
              "setter": {
                "kind": "literal",
                "literal": {
                  "kind": "string",
                  "value": "First Last",
                },
              },
            },
            {
              "kind": "basic",
              "name": "prefix",
              "setter": {
                "kind": "literal",
                "literal": {
                  "kind": "string",
                  "value": "Mr. ",
                },
              },
            },
          ],
          "hook": {
            "file": "hooks2.js",
            "kind": "source",
            "runtimeName": "TextRuntime",
            "target": "prefix",
          },
        },
        "kind": "hook",
      },
    },
    {
      "kind": "basic",
      "name": "avg",
      "setter": {
        "hook": {
          "args": [
            {
              "kind": "basic",
              "name": "x",
              "setter": {
                "kind": "literal",
                "literal": {
                  "kind": "integer",
                  "value": 100,
                },
              },
            },
            {
              "kind": "basic",
              "name": "y",
              "setter": {
                "kind": "literal",
                "literal": {
                  "kind": "integer",
                  "value": 20,
                },
              },
            },
          ],
          "hook": {
            "file": "hooks.js",
            "kind": "source",
            "runtimeName": "MathRuntime",
            "target": "prefix",
          },
        },
        "kind": "hook",
      },
    },
  ],
  "isPrimary": true,
  "kind": "create-one",
  "model": "Result",
  "select": [
    {
      "alias": "id",
      "expr": {
        "kind": "identifier-path",
        "namePath": [
          "$action_0",
          "id",
        ],
      },
      "kind": "expression",
      "type": {
        "kind": "integer",
        "nullable": false,
      },
    },
  ],
  "targetPath": [
    "Result",
  ],
}
`;
